src/shogun/lib/DynamicArray.h

```diff
@@ -654,6 +802,876 @@ template <class T> class CDynamicArray :public CSGObject
+				    dim1_size, dim2_size, dim3_size)
+			else
+				SG_PRINT(
+				    "DynamicArray of size: %dx%dx%d\n", dim1_size, dim2_size,
+				    dim3_size)
+
+			for (int32_t k = 0; k < dim3_size; k++)
+				for (int32_t j = 0; j < dim2_size; j++)
+				{
+					SG_PRINT("element(%d,:,%d) = [ ", j, k)
+					for (int32_t i = 0; i < get_num_elements(); i++)
+						SG_PRINT("%1.1f,", (float32_t)element(i, j, k))
+					SG_PRINT(" ]\n")
+				}
+		}
+
+		/** display array's size */
+		inline void display_size()
+		{
+			SG_PRINT(
+			    "DynamicArray of size: %dx%dx%d\n", dim1_size, dim2_size,
+			    dim3_size)
+		}
+
+		/** @return object name */
+		virtual const char* get_name() const
+		{
+			return "DynamicArray";
+		}
+
+		/** Can (optionally) be overridden to pre-initialize some member
+		 *  variables which are not PARAMETER::ADD'ed.  Make sure that at
+		 *  first the overridden method BASE_CLASS::LOAD_SERIALIZABLE_PRE
+		 *  is called.
+		 *
+		 *  @exception ShogunException Will be thrown if an error
+		 *                             occurres.
+		 */
+		virtual void load_serializable_pre() throw(ShogunException)
+		{
+			CSGObject::load_serializable_pre();
+
+			resize_array(get_num_elements(), true);
+		}
+
+		/** Can (optionally) be overridden to pre-initialize some member
+		 *  variables which are not PARAMETER::ADD'ed.  Make sure that at
+		 *  first the overridden method BASE_CLASS::SAVE_SERIALIZABLE_PRE
+		 *  is called.
+		 *
+		 *  @exception ShogunException Will be thrown if an error
+		 *                             occurres.
+		 */
+		virtual void save_serializable_pre() throw(ShogunException)
+		{
+			CSGObject::save_serializable_pre();
+			resize_array(get_num_elements(), true);
+		}
+
+		virtual CSGObject* clone()
+		{
+			CDynamicArray* cloned = (CDynamicArray*)CSGObject::clone();
+			// Since the array vector is registered with
+			// current_num_elements as size (see parameter
+			// registration) the cloned version has less memory
+			// allocated than known to dynarray. We fix this here.
+			// cloned->num_elements = cloned->array_size;
+			return cloned;
+		}
+
+	private:
+		/** register parameters */
+		virtual void init()
+		{
+			set_generic<bool>();
+			m_parameters->add_vector(
+			    &m_array, &num_elements, "array", "Memory for dynamic array.");
+
+			SG_ADD(
+			    &free_array, "free_array", "whether array must be freed",
+			    MS_NOT_AVAILABLE);
+			SG_ADD(&dim1_size, "dim1_size", "Dimension 1", MS_NOT_AVAILABLE);
+			SG_ADD(&dim2_size, "dim2_size", "Dimension 2", MS_NOT_AVAILABLE);
+			SG_ADD(&dim3_size, "dim3_size", "Dimension 3", MS_NOT_AVAILABLE);
+		}
+
+	protected:
+		/** underlying array */
+		bool* m_array;
+
+		/** number of elements */
+		int32_t num_elements;
+
+		/** dimension 1 */
+		int32_t dim1_size;
+
+		/** dimension 2 */
+		int32_t dim2_size;
+
+		/** dimension 3 */
+		int32_t dim3_size;
+
+		/** the number of currently used elements */
+		int32_t array_size;
```

Comment:
The array_size in here will indicate the size of this array like how many memory been allocated to this array(like `m_array.size() == array_size`). And `num_elements` means how many elements be used in this array. So they are not overlap actually.

Sorry for the wrong conclusion I made last time.